documentation/high-level-introduction.md

```diff
@@ -0,0 +1,257 @@
+  - `ToolTask` - A base class that makes invoking a single external tool significantly easier
+- Parameters are passed via public properties defined on the task
+
+```c#
+public abstract class MySpecialTask : Task 
+{
+    public string MyOptionalArg { get; set; }
+    
+    [Required]
+    public string MyRequiredArg { get; set; }
+}
+```
+
+  - Properties can be marked [Required] which will fail the build if given an empty string value
+  - Strings are the common type, but you can use types like bool and msbuild will convert for you
+  - ITaskItem[] are need to pass Items (lists) to \ from tasks when the task may need to modify them
+- They are later invoked with:
+
+```xml
+     <UsingTask TaskName="MyTasks.MySpecialTask" AssemblyFile="MyTasks.dll" />
+     
+    <Target Name="_MySpecialTask">
+                    <MySpecialTask
+                            MyOptionalArg="$(MyOptionalArg)"
+                            MyRequiredArg="$(MyRequiredArg)"
+                            >
+                    </MySpecialTask>
+    </Target>
+```
+- You can attach a C# debugger to the task, but that can be tricky on mono since there is no attach to process
+- `SessionId` is needed on most C# tasks to enable “build remoting” for Visual Studio Windows.
+
+## Execution Model
+
+A very inaccurate but useful way of thinking about how msbuild processes is to split it into three parts.
+
+- Loading
+  - Starting with the csproj load each file included into a giant buffer
+    - Think C preprocessor gloming files together
+  - Some default values will be set in the global property dictionary
+- Property Eval
+  - Starting from the top, evaluate every node
+    - Property and Item Group will set items in the global context
+    - UsingTask will load task assemblies and prepare to invoke them
+    - Targets will define then and their ordering\inputs\output
+      - **Their contents will not be executed yet**
+  - This will respect conditions, and be done top to bottom based upon include order
+- Target Eval
+  - Starting with the defined “DefaultTargets", which for our purposes is always “Build” determine a list of target to invoke and their ordering
+  - Invoke one at a time, comparing their Input\Output against file timestamps like make
+
+
+## "SDK style” project and nuget injection
+  - SDK style projects and nugets are beyond the scope of this document 
+  - One important thing to note is the “magic” injection of files
+    - When referenced the props file will be “auto injected” to the top of your project and the targets injected at the very bottom. 
+    - Thus you can include files for these project types you don’t explicitly see referenced in the csproj
+
+
+## Debugging
+  - Easily the most important thing is to get a full diagnostic build. This will list all variables set and the ordering\execution of each target.
+  - You can add debug “Console.WriteLine”s by using the built in MessageTask:
+
+
```

Comment:
Extra line?

Also, should it be mentioned that if you're trying to change MSBuild itself, you need to make the messages verbose?